### PR TITLE
Pass graph to parallel executor instead of program

### PR DESCRIPTION
**Pull request by panyx0718**
sometimes parallel executor needs to run graphs because there are some pass before parallel executor

### CODE REVIEW

File: paddle/fluid/framework/parallel_executor.cc [hunk 1]
```diff
@@ -265,33 +267,37 @@ ParallelExecutor::ParallelExecutor(
...
   if (build_strategy.enable_parallel_graph_) {
     for (size_t i = 0; i < member_->places_.size(); ++i) {
-      std::unique_ptr<ir::Graph> graph = build_strategy.Apply(
-          main_program, {member_->places_[i]}, loss_var_name,
+      temp_owned_graph = build_strategy.Apply(
+          std::move(temp_owned_graph), {member_->places_[i]}, loss_var_name,
           {member_->local_scopes_[i]}, member_->nranks_, member_->use_cuda_,
           member_->nccl_ctxs_.get());
-      graphs.push_back(std::move(graph));
+      graphs.push_back(temp_owned_graph.release());
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc]: `temp_owned_graph.release()` will return a pointer to the managed object and releases the ownership. It should not be released more than one time.

**Review by chengduoZH — DISMISSED** [review]

**Review by chengduoZH — DISMISSED** [review]

File: python/paddle/fluid/contrib/slim/unitest/test_quantization_pass.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+# you may obtain a copy of the license at
+#
+#     http://www.apache.org/licenses/license-2.0
+#
+# unless required by applicable law or agreed to in writing, software
+# distributed under the license is distributed on an "as is" basis,
+# without warranties or conditions of any kind, either express or implied.
+# see the license for the specific language governing permissions and
+# limitations under the license.
+
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/unitest/test_quantization_pass.py]: `test_quantization_pass.py` has been moved in the `python/paddle/fluid/contrib/slim/tests` dir. If this is not necessary， please remove it.

File: python/paddle/fluid/framework.py [hunk 3]
```diff
@@ -2332,6 +2332,7 @@ def _construct_from_desc(desc):
         """
         p = Program()
         p.desc = desc
+        # TODO(wangzhen): Block.vars/ops are not filled, should fix it.
```
Comment (wzzju) [inline python/paddle/fluid/framework.py]: Block.vars/ops have been filled. Thx.

File: python/paddle/fluid/compiler.py [hunk 4]
```diff
@@ -101,6 +116,7 @@ def with_data_parallel(self,
             self
         """
         assert not self._is_data_parallel, "Already compiled with parallel."
+        assert not self._is_inference, "Cannot compile both data parallel and inference"
```
Comment (chengduoZH) [inline python/paddle/fluid/compiler.py:119]: The error doc is confusing, it should be `with_data_parallel and with_inference_optimize cannot be used together.`

Comment (panyx0718) [reply python/paddle/fluid/compiler.py:119]: done

**Review by chengduoZH — APPROVED** [review]

**Pull request merged**